[SPARK-8538][SPARK-8539][ML] Linear Regression Training and Testing Results

feynmanliang opened the pull request:
Adds results (e.g. objective value at each iteration, residuals) on training and user-specified test sets for LinearRegressionModel.

Notes to Reviewers:
- Are the `*TrainingResults` and `Results` classes too specialized for `LinearRegressionModel`? Where would be an appropriate level of abstraction?
- Please check `@transient` annotations are correct; the datasets should not be copied and kept during serialization.
- Any thoughts on `RDD`s versus `DataFrame`s? If using `DataFrame`s, suggested schemas for each intermediate step? Also, how to create a "local DataFrame" without a `sqlContext`?

Commit:
Add training results and model summary to ML LinearRegression

Commit:
Use RDD[(Double, Double)] instead of DF

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36034 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36034/consoleFull) for   PR 7099 at commit [`521f397`](https://github.com/apache/spark/commit/521f3974db669d6d5b0ecf29b688e888105f858e).

SparkQA commented:
[Test build #36034 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36034/console) for   PR 7099 at commit [`521f397`](https://github.com/apache/spark/commit/521f3974db669d6d5b0ecf29b688e888105f858e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LinearRegressionTrainingResults(`
  * `class LinearRegressionResults(@transient val predictionAndObservations: RDD[(Double, Double)])`
  * `class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Style guide fixes

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36036 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36036/consoleFull) for   PR 7099 at commit [`dc51bce`](https://github.com/apache/spark/commit/dc51bce2f95468975e9ef0fb807d8a18feb2df8c).

Commit:
Add LinearRegressionModel.evaluate() to get results on test sets

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36041/consoleFull) for   PR 7099 at commit [`97b0a81`](https://github.com/apache/spark/commit/97b0a81d8ee02acf6b2b684be73ea29d75d7bd8c).

SparkQA commented:
[Test build #36036 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36036/console) for   PR 7099 at commit [`dc51bce`](https://github.com/apache/spark/commit/dc51bce2f95468975e9ef0fb807d8a18feb2df8c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LinearRegressionTrainingResults(`
  * `class LinearRegressionResults(@transient val predictionAndObservations: RDD[(Double, Double)])`
  * `class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #36041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36041/console) for   PR 7099 at commit [`97b0a81`](https://github.com/apache/spark/commit/97b0a81d8ee02acf6b2b684be73ea29d75d7bd8c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LinearRegressionTrainingResults(`
  * `class LinearRegressionResults(@transient val predictionAndObservations: RDD[(Double, Double)])`
  * `class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Add LinearRegressionTrainingResults tests, make test suite code copy+pasteable

Commit:
Add test for decreasing objective trace

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #36050 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36050/consoleFull) for   PR 7099 at commit [`1cedb2b`](https://github.com/apache/spark/commit/1cedb2beb41935ef462627d4c055c06a38573ee3).

SparkQA commented:
[Test build #36050 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36050/console) for   PR 7099 at commit [`1cedb2b`](https://github.com/apache/spark/commit/1cedb2beb41935ef462627d4c055c06a38573ee3).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `class LinearRegressionTrainingResults(`
  * `class LinearRegressionResults(@transient val predictionAndObservations: RDD[(Double, Double)])`
  * `class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)])`

AmplabJenkins commented:
Build finished. Test PASSed.

Commit:
Merge branch 'master' into SPARK-8538

* master: (86 commits)
  [SPARK-8236] [SQL] misc functions: crc32
  [SPARK-8680] [SQL] Slightly improve PropagateTypes
  [SPARK-8723] [SQL] improve divide and remainder code gen
  [SPARK-8590] [SQL] add code gen for ExtractValue
  [SPARK-7756] [CORE] More robust SSL options processing.
  [SPARK-8551] [ML] Elastic net python code example
  [SPARK-8434][SQL]Add a "pretty" parameter to the "show" method to display long strings
  [SPARK-5161] [HOTFIX] Fix bug in Python test failure reporting
  [SPARK-8650] [SQL] Use the user-specified app name priority in SparkSQLCLIDriver or HiveThriftServer2
  [SPARK-8721][SQL] Rename ExpectsInputTypes => AutoCastInputTypes.
  MAINTENANCE: Automated closing of pull requests.
  [SPARK-5161] Parallelize Python test execution
  [SPARK-7667] [MLLIB] MLlib Python API consistency check
  [SPARK-8669] [SQL] Fix crash with BINARY (ENUM) fields with Parquet 1.7
  [SPARK-8715] ArrayOutOfBoundsException fixed for DataFrameStatSuite.crosstab
  [SPARK-8456] [ML] Ngram featurizer python
  Revert "[SPARK-8437] [DOCS] Using directory path without wildcard for filename slow for large number of files with wholeTextFiles and binaryFiles"
  [SPARK-8019] [SPARKR] Support SparkR spawning worker R processes with a command other then Rscript
  [SPARK-8410] [SPARK-8475] remove previous ivy resolution when using spark-submit
  [SPARK-8437] [DOCS] Using directory path without wildcard for filename slow for large number of files with wholeTextFiles and binaryFiles
  ...

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36160 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36160/consoleFull) for   PR 7099 at commit [`ffceaec`](https://github.com/apache/spark/commit/ffceaecff6db525855ab2a23e8e2326e5a0ed523).

SparkQA commented:
[Test build #36160 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36160/console) for   PR 7099 at commit [`ffceaec`](https://github.com/apache/spark/commit/ffceaecff6db525855ab2a23e8e2326e5a0ed523).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class LinearRegressionTrainingResults(`
  * `class LinearRegressionResults(@transient val predictionAndObservations: RDD[(Double, Double)])`
  * `class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)])`
  * `trait ExtractValue `
  * `abstract class ExtractValueWithStruct extends UnaryExpression with ExtractValue `
  * `abstract class ExtractValueWithOrdinal extends BinaryExpression with ExtractValue `
  * `case class Crc32(child: Expression)`

AmplabJenkins commented:
Merged build finished. Test FAILed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -140,7 +143,10 @@ class LinearRegression(override val uid: String)
-      return new LinearRegressionModel(uid, Vectors.sparse(numFeatures, Seq()), yMean)
+      val weights = Vectors.sparse(numFeatures, Seq())
+      val intercept = yMean
+      val summary = generateTrainingResults(instances, Array(), weights, intercept)
```
Is this a WIP?  (objectiveTrace and weights)

feynmanliang replied:
Nope. If all labels are the same there is no objectiveTrace and all the weights are zero (since just predicting `yMean` = the single correct label will give 0 loss).

jkbradley replied:
Gotcha.  (I should have looked more at the context.)

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -193,11 +199,36 @@ class LinearRegression(override val uid: String)
+    copyValues(new LinearRegressionModel(uid, weights.compressed, intercept, summary))
+  }
+
+  private def generateTrainingResults(
```
I'd prefer this method be in LinearRegressionTrainingResults, either as a constructor, or as a static factory method.

Also, you might want to pass in the LinearRegressionModel so that you can use its transform() method.  Rather than having LinearRegressionModel take summary in the constructor, it can start with it set to null; later on, you can set it using a setter method.

feynmanliang replied:
OK.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -208,20 +239,61 @@ class LinearRegression(override val uid: String)
+ * @param objectiveTrace objective function value at each iteration.
+ */
+@Experimental
+class LinearRegressionTrainingResults(
```
We should keep this constructor private for now

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -208,20 +239,61 @@ class LinearRegression(override val uid: String)
+ * @param predictionAndObservations an RDD of (prediction, observation) pairs.
+ */
+@Experimental
+class LinearRegressionResults(@transient val predictionAndObservations: RDD[(Double, Double)])
```
We should keep this constructor private for now

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -208,20 +239,61 @@ class LinearRegression(override val uid: String)
+ */
+@Experimental
+class LinearRegressionResults(@transient val predictionAndObservations: RDD[(Double, Double)])
+  extends RegressionMetrics(predictionAndObservations) {
```
I like the code-sharing motivation behind extending RegressionMetrics, but we should not include this dependency between the ml and mllib packages.  We can use RegressionMetrics internally, but we should create a copy of each of the methods in this class.  (Or those methods could be in a private abstraction which can be shared for later regression models.)

jkbradley commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala:
```diff
@@ -31,7 +31,8 @@ import org.apache.spark.sql.DataFrame
 @Experimental
-class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)]) extends Logging {
+class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)])
+  extends Logging with Serializable {
```
Does this have to be Serializable?

If so, I'm not quite sure how extending Serializable interacts with the lazy val.  It would be useful to know when that lazy val gets instantiated (hopefully just once when the object is serialized).

feynmanliang replied:
Not too sure why but I was getting `Task Not Serializable` errors without this. I suspect this is because everything inside the `train()` method's closure gets serialized.

I followed up on how lazy vals interact with serialization and found [this SO post](http://stackoverflow.com/questions/27882307/how-does-serialization-of-lazy-fields-work) which says that the value is serialized iff it was computed before serialization.

In my updated implementation, one option could be to force the evaluation of `RegressionMetrics.summary` in the `LinearRegressionTestResults` constructor. However, despite being serializable I don't expect this class to be replicated anywhere except at the driver so maybe this eager evaluation is unnecessary. @jkbradley thoughts?

jkbradley replied:
I think the serialization errors are from LinearRegressionModel being used in closures (for prediction).  That should be fixable by having trainingResults be transient.  We'll have to add doc to the getter/setter for it to warn the user about using those only on the driver.

I'd strongly prefer lazy evaluation.

feynmanliang replied:
OK.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -208,20 +239,61 @@ class LinearRegression(override val uid: String)
+ */
+@Experimental
+class LinearRegressionTrainingResults(
+    @transient predictionAndObservations: RDD[(Double, Double)],
```
No need to mark this transient, I believe, since it's not stored as a val

feynmanliang replied:
OK

jkbradley commented:
> Are the *TrainingResults and Results classes too specialized for LinearRegressionModel? Where would be an appropriate level of abstraction?

It's OK if we add abstractions now or later on, and those abstractions can be private at first if we are uncertain about the API.  The public methods and classes won't be able to change in the future, so we do need to think about those for sure.

> Any thoughts on RDDs versus DataFrames? If using DataFrames, suggested schemas for each intermediate step? Also, how to create a "local DataFrame" without a sqlContext?

I think we should try to use DataFrames instead of RDDs wherever possible within the spark.ml APIs.  Hopefully the schema will be clear based on the source of the data (e.g., following the input and output schema of the model.transform method).

You can create a SQLContext from the given DataFrame's SparkContext.

Commit:
Refactored LinearRegressionResults API to be more private

feynmanliang commented:
It wasn't clear to me how to use DFs in the result classes I was creating since they didn't have access to the model parameters (featuresCol, predictionCol, etc). I could add them as constructor params if you think that'll be better but it's not clear to me what the benefit of using a DF in the result classes is since most use cases will only be interested in the summary functions rather than the predictions + labels themselves.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
Fix imports

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36295 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36295/consoleFull) for   PR 7099 at commit [`9509c79`](https://github.com/apache/spark/commit/9509c799347738dc58d104ef5c4ae68f4c275c84).

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #36295 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36295/console) for   PR 7099 at commit [`9509c79`](https://github.com/apache/spark/commit/9509c799347738dc58d104ef5c4ae68f4c275c84).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

jkbradley commented:
> It wasn't clear to me how to use DFs in the result classes I was creating since they didn't have access to the model parameters (featuresCol, predictionCol, etc). I could add them as constructor params if you think that'll be better but it's not clear to me what the benefit of using a DF in the result classes is since most use cases will only be interested in the summary functions rather than the predictions + labels themselves.

Sorry, this was ambiguous.  The plan is to have DFs for each result type, not necessarily ones zipped with the transformed data.  Later on, we could provide extra output columns to include the values in the transformed data, but we won't just yet.  E.g., we can provide a DataFrame storing only 1 column of residuals.

Commit:
Use DFs and localize serialization to LinearRegressionModel

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36475 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36475/consoleFull) for   PR 7099 at commit [`509ae36`](https://github.com/apache/spark/commit/509ae360a539ca8e3d2906c8951d017ef2fc0627).

SparkQA commented:
[Test build #36475 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36475/console) for   PR 7099 at commit [`509ae36`](https://github.com/apache/spark/commit/509ae360a539ca8e3d2906c8951d017ef2fc0627).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36480 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36480/consoleFull) for   PR 7099 at commit [`072e948`](https://github.com/apache/spark/commit/072e9484eb2750952340ceb10d553dfac6768471).

SparkQA commented:
[Test build #36480 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36480/console) for   PR 7099 at commit [`072e948`](https://github.com/apache/spark/commit/072e9484eb2750952340ceb10d553dfac6768471).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

jkbradley commented:
Reviewing now

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -194,10 +203,15 @@ class LinearRegression(override val uid: String)
+    val model = new LinearRegressionModel(uid, weights.compressed, intercept)
+    val trainingResults = new LinearRegressionTrainingResults(
+      model.transform(dataset).select(model.getPredictionCol, model.getLabelCol),
+      lossHistory.result())
```
I'm pinging DB about this issue: [https://issues.apache.org/jira/browse/SPARK-8845]

jkbradley replied:
He's fixing this, so don't worry about this issue for now.

dbtsai replied:
lossHistory in LiR is in the scaled space, so we have to transformed back. Let me do this. For LOR, loglikliehood will not be changed under feature transformation.

dbtsai replied:
Okay, lossHistory will still be the one returned in the scaled space. It's not easy and meaningful to convert it into original space.

feynmanliang replied:
Okay. I will leave as is and document that LossHistory is for debugging purposes and not actual negative log likelihood

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +226,110 @@ class LinearRegressionModel private[ml] (
+   * Gets results (e.g. residuals, mse, r^2) of model on training set. This method should only
+   * be called on the driver (it is not available on workers).
+   */
+  def getTrainingResults: Option[LinearRegressionTrainingResults] = trainingResults
```
Let's stick with the design doc naming, or modify the design doc if you think other naming is better.  I like the more succinct usage in the doc:
```
lrModel.summary.objectiveTrace
```
That will require summary to throw an exception if the results are not available, but that seems OK to me as long as we make it a meaningful exception.  We could also provide:
```
def hasSummary: Boolean
def setSummary(summary: LinearRegressionSummary): this.type
def getSummary // for convenience, if you want
```

I also like calling it Summary instead of TrainingResults since it's shorter and may be more recognizable to R users.

feynmanliang replied:
OK, will change `trainingResults` to `summary`.

wrt results not being available and `hasSummary`, I'm currently using an `Option` to handle those two cases. Would you suggest something else?

jkbradley replied:
Option isn't very Java-friendly, so providing ```summary``` and ```hasSummary``` might be best.  You could keep ```getSummary``` as the Scala version which returns an Option.

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +226,110 @@ class LinearRegressionModel private[ml] (
+  }
+
+  /**
+   * Evaluates the model on a test-set.
```
no hyphen

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +226,110 @@ class LinearRegressionModel private[ml] (
+
+  /**
+   * Evaluates the model on a test-set.
+   * @param testset Test dataset to evaluate model on.
```
Let's just call it "dataset" for uniformity

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +226,110 @@ class LinearRegressionModel private[ml] (
+  /**
+   * Evaluates the model on a test-set.
+   * @param testset Test dataset to evaluate model on.
+   * @return
```
Remove, or fill in.

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +226,110 @@ class LinearRegressionModel private[ml] (
+@Experimental
+class LinearRegressionTrainingResults private[ml] (
+    predictionAndLabel: DataFrame,
+    val objectiveTrace: Array[Double])
```
This should be a local DataFrame constructed from a Seq, with 2 columns: iteration, objective

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +226,110 @@ class LinearRegressionModel private[ml] (
+/**
+ * :: Experimental ::
+ * Linear regression results evaluated on a dataset.
+ * @param predictionAndLabel dataframe with columns prediction(0) and label (1).
```
If we're providing part of the transformed data, let's provide the whole transformed dataset with all of the columns.  We can call it ```predictions: DataFrame``` and have it be the output of model.transform(data).

Also: In general, it's best to refer to columns by name.  I'm not sure if there are strong guarantees about ordering of column names, except when specified via a select statement.  I.e., if you call ```df.select("a", "b")``` then you know the column ordering of the selected data.  Otherwise, I would not make assumptions.

feynmanliang replied:
There are no guarantees on column names and `inputCol`/`outputCol` are currently not visible from within `LinearRegressionSummary`, hence the numeric indexing and restriction on columns.

I am considering another design mixing in `LinearRegressionSummary` as a trait with a self-annotation for `HasInputCol`, which would give me access to `inputCol` as well as push more correctness checking to the type-checker.

jkbradley replied:
Right, that makes sense, and I had not thought about it properly before.  I think it would be reasonable for LinearRegressionResults to inherit from the needed input/output column Params.

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +226,110 @@ class LinearRegressionModel private[ml] (
+   */
+  def r2: Double = metrics.r2
+
+  /** Get residuals */
```
Define residuals in doc to make it clear that it's pred - label (not vice versa).

feynmanliang replied:
OK.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala:
```diff
@@ -87,6 +86,42 @@ class LinearRegressionSuite extends SparkFunSuite with MLlibTestSparkContext {
+    }.zip(model.getTrainingResults.get.residuals.map(_.getDouble(0)))
+      .collect()
+      .foreach {
+      case (manualResidual: Double, resultResidual: Double) =>
```
style: put case on previous line after brace

Actually, there is a comparison for Array, so it'd be simpler to use that

feynmanliang replied:
Which comparison can be used for elementwise approximate equals with tolerance (I'm comparing two `Array[Double]`s)? I saw `IndexSeqOptimized#sameElements` but that tests for exact equality (`==`) and `TestingUtils` seems to only have stuff for `Vectors`.

jkbradley replied:
Yeah, I was probably recalling Vector, not Array.  Please ignore

jkbradley commented:
The unit test should test evaluate() as well.

jkbradley commented:
That's all for now!

Commit:
Refactor from jkbradley code review

Commit:
Add Params to LinearModel*Summary, refactor tests and add test for evaluate()

Commit:
Optimize imports

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36614 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36614/consoleFull) for   PR 7099 at commit [`c2fe835`](https://github.com/apache/spark/commit/c2fe835e7d5c2fa9893537d35386e9f93ddf9306).

SparkQA commented:
[Test build #36614 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36614/console) for   PR 7099 at commit [`c2fe835`](https://github.com/apache/spark/commit/c2fe835e7d5c2fa9893537d35386e9f93ddf9306).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Replace Params from LinearRegression*Summary with private constructor vals

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36623/consoleFull) for   PR 7099 at commit [`0a605d8`](https://github.com/apache/spark/commit/0a605d859d96797d2511127dcb89b104ff595330).

SparkQA commented:
[Test build #36623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36623/console) for   PR 7099 at commit [`0a605d8`](https://github.com/apache/spark/commit/0a605d859d96797d2511127dcb89b104ff595330).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -140,7 +142,16 @@ class LinearRegression(override val uid: String)
+        $(predictionCol),
+        $(labelCol),
+        Array(0D))
+      return model.setSummary(trainingSummary)
```
```scala
return copyValues(model.setSummary(trainingSummary))
```

feynmanliang replied:
OK

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +230,139 @@ class LinearRegressionModel private[ml] (
+        new NullPointerException())
+  }
+
+  def setSummary(summary: LinearRegressionTrainingSummary): this.type = {
```
Do we want to make it private to this scope?

jkbradley replied:
Good point, I agree

feynmanliang replied:
OK.

jkbradley commented:
@dbtsai Good point about the objective being in the scaled space.  If Breeze will give us separate loss and regularization values, couldn't we compute the non-scaled objective by hand in LinearRegression efficiently?  We could rescale the loss f(x) using labelStd, and the regularization term should already be OK (not really scaled).  Does that sound feasible?

dbtsai commented:
In general, the losses returned from optimizer are for debugging purpose, and users can see how well the optimizer works. The reason is the real statistical metrics users care about are RSE or loglikihood which can be used for computing other metrics don't involve the regTerm. However, the returned loss like the following has regTerm, which is not the real RSE. Also, no easy/cheap way to convert back to the loss which standardization. As a result, we can have the lossHistroy for debugging purpose, and have another pass of data to compute RSE and loglikihood.  

```latex
L = 1/2n ||\sum_i w_i(x_i - \bar{x_i}) / \hat{x_i} - (y - \bar{y}) / \hat{y}||^2 + regTerms
```

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +230,139 @@ class LinearRegressionModel private[ml] (
+      throw new SparkException(
+        "No training summary available for this LinearRegressionModel",
+        new NullPointerException())
+  }
```
I have concerns to throw out the exception. Since the whole `LinearRegressionTrainingSummary` is transient, users may mistakenly terminate the program if the models are loaded from files instead of created in training. Thus, we have to ask users to be very carefully in reusing the code in training and online system.

feynmanliang replied:
OK. See proposed changes to make predictions dataframe transient.

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +230,139 @@ class LinearRegressionModel private[ml] (
   extends RegressionModel[Vector, LinearRegressionModel]
   with LinearRegressionParams {
 
+  @transient private var trainingSummary: Option[LinearRegressionTrainingSummary] = None
```
If we make the whole thing `transient`, does it mean even the small object like `lossHistory` can not be saved as part of model? I'm still leaning toward having lossHistory as part of LinearRegressionModel, and link to training dataframe using @transient. In fact, in R, those small stats are part of models.

feynmanliang replied:
OK. See proposed changes to make predictions dataframe transient.

jkbradley replied:
@dbtsai I would prefer to have stats included within the same object, rather than having separate objects for small (local) stats vs. large (distributed) stats.  Do you have a good use case for when a worker would need to access these stats?  Even in that case, the driver could get the needed stats and then pass those to the worker.

feynmanliang replied:
What about during model save/load? Users may want to see some training stats such as r^2 and MSE after loading a model. If `trainingSummary` is transient and the only place we put the small stats, we will need to think of another way to support this use case.

jkbradley replied:
Per our discussions today, this should no longer be transient.

jkbradley commented:
It's odd to me that there is not a natural conversion from scaled loss to unscaled loss; the scaled loss should essentially be a loss scaled by the mean label, which we know.  The main issue is getting separate loss and regularization terms back from Breeze.

But I'm fine with providing only the scaled loss.  @feynmanliang Can you please include a comment in the Scala doc about that?

Btw, I'm distinguishing "loss" (e.g., squared error) from "objective" (e.g., squared error + regularization), and I hope we can stick with this common usage throughout documentation.

feynmanliang commented:
It seems like `objectiveTrace` is going to return scaled loss + regularization since #7245 changes Breeze to return `adjustedValue`

Commit:
Make TrainingSummary and metrics serializable, prediction dataframe transient

feynmanliang commented:
Please see new commits addressing comments changing `TrainingSummary` to be non-transient and pushing transient down to `prediction : DataFrame`. In particular, please review the new `Serializable` classes and the eager evaluation of the lazy RegressionMetrics.summary object.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36705 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36705/consoleFull) for   PR 7099 at commit [`a65dfda`](https://github.com/apache/spark/commit/a65dfdab22ada51160831b1915c112458b05f9cc).

Commit:
Fix failing Java test

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +230,140 @@ class LinearRegressionModel private[ml] (
+  )
+
+  // Force evaluation of lazy RegressionMetrics.summary for proper serialization
+  metrics.explainedVariance
```
Will this suffice to ensure that metrics serializes correctly even though predictions are transient? Should we consider just making RegressionMetrics.summary not lazy?

jkbradley replied:
We should not force evaluation.  This will only be needed if the LinearRegressionSummary gets serialized and sent to workers for them to use the metrics, but I had planned not to support that.  Users can still extract metrics on the driver and then send those values to workers.  Why does LinearRegressionSummary need to be serializable?

feynmanliang replied:
I agree wtih @dbtsai that not all of `LinearRegressionSummary` should be transient since small things (objectiveTrace, even the metrics like r^2, MSE, etc) can be kept with the `LinearRegressionTrainingSummary` of each `LinearRegressionModel`.

Since `LinearRegressionModel` is sent to workers, all its fields need to be serializable.

jkbradley replied:
This is going to make the API very confusing.  The summary will contain many member vals, some of which are transient and some of which are not.  Even worse, some simple vals like Double and Int will be transient.  To fix this, we would need to force eager evaluation, which is not acceptable.  (It is reasonably cheap now, but it will become more expensive if we provide other more costly stats, such as medians and quantiles.)

jkbradley replied:
I have a hard time thinking of why a worker needs to have the objective trace; that's the sort of thing people would just want to print from the driver.  I'm sure that, at some point, someone will need workers to access some stats, but those can be extracted on the driver & then passed to the workers.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala:
```diff
@@ -31,7 +31,8 @@ import org.apache.spark.sql.DataFrame
 @Experimental
-class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)]) extends Logging {
+class RegressionMetrics(@transient predictionAndObservations: RDD[(Double, Double)])
+  extends Logging with Serializable {
```
Should we consider making summary no longer lazy now that this is serializable?

jkbradley replied:
I still don't think this should be serializable.

feynmanliang replied:
I agree that `RegressionMetrics` will probably not need to be serialized and sent to workers. However, `LinearRegressionModel` needs to be serializable => `LinearRegressionSummary` and its fields need to as well.

jkbradley replied:
I see, I misunderstood yall's comments.  Does that mean you couldn't get it to work just by marking the summary field in the model transient?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36705 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36705/console) for   PR 7099 at commit [`a65dfda`](https://github.com/apache/spark/commit/a65dfdab22ada51160831b1915c112458b05f9cc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegressionMetrics(@transient predictionAndObservations: RDD[(Double, Double)])`

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #36707 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36707/consoleFull) for   PR 7099 at commit [`1d9ea42`](https://github.com/apache/spark/commit/1d9ea42048b6e54e4b84514608030b2f6a2022ea).

SparkQA commented:
[Test build #36707 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36707/console) for   PR 7099 at commit [`1d9ea42`](https://github.com/apache/spark/commit/1d9ea42048b6e54e4b84514608030b2f6a2022ea).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RegressionMetrics(@transient predictionAndObservations: RDD[(Double, Double)])`

AmplabJenkins commented:
Merged build finished. Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -212,12 +230,140 @@ class LinearRegressionModel private[ml] (
   extends RegressionModel[Vector, LinearRegressionModel]
   with LinearRegressionParams {
 
+  private var trainingSummary: Option[LinearRegressionTrainingSummary] = None
```
To make sure LinearRegressionModel is serializable, can we not just make this transient?

feynmanliang replied:
That works, and it's what I had in an earlier PR, but see discussion above (TLDR: r^2, MSE and other small stats might be useful to have especially for model save/load, marking this transient will make those stats inaccessible).

jkbradley replied:
Btw, now that there is no public setter, let's make this a val and pass it to a private constructor.

jkbradley replied:
I see; I thought the use case was for sending stats to workers.  For model save/load, I agree we will need to save local stats by default.  I'm not sure that marking items as transient is the way to do that, unless we want to encourage users to save models as POJOs (which we should not).  I think that a follow-up PR should add the summary to the saved model, where it saves only the local values.  When the model is loaded from disk, the values not saved can be set to null.  Is there a need to mark things as transient?

jkbradley commented:
@dbtsai After discussing with @feynmanliang I think we have a plan.  Here's the breakdown of the issues with transient & serialization:

Serialization could be important for:
(1) Java serialization for persisting models & stats
(2) sending models & stats to workers

Marking some things as transient should fix these issues, at least partly.  Assuming we mark enough data members as transient, then these 2 issues are answered as follows:

(1) Java serialization:
* Users should be able to save models as POJOs and load them back.  However, stats will not be saved.
* We can later add support for saving stats to our native model save/load.

(2) sending to workers:
* We can send models to workers for predictions.
* We will not assume stats are available on workers.

Given these decisions, it suffices to mark the summary data member stored in LinearRegressionModel as transient, and not making the results & metrics classes Serializable.

Later on, we could modify what is transient, or make more things Serializable.  (Changing transient modifiers does not break binary compatibility.)

@dbtsai Does this summary seem reasonable?

Commit:
Make TrainingSummary transient and remove Serializable from *Summary and RegressionMetrics

Commit:
Merge branch 'master' into SPARK-8538

* master: (130 commits)
  [SPARK-8704] [ML] [PySpark] Add missing methods in StandardScaler
  [SPARK-8559] [MLLIB] Support Association Rule Generation
  [SPARK-8821] [EC2] Switched to binary mode for file reading
  [SPARK-8823] [MLLIB] [PYSPARK] Optimizations for SparseVector dot products
  [SPARK-8711] [ML] Add additional methods to PySpark ML tree models
  [SPARK-8570] [MLLIB] [DOCS] Improve MLlib Local Matrix Documentation.
  [SPARK-8788] [ML] Add Java unit test for PCA transformer
  [SPARK-6731] [CORE] Addendum: Upgrade Apache commons-math3 to 3.4.1
  [HOTFIX] Rename release-profile to release
  [SPARK-8759][SQL] add default eval to binary and unary expression according to default behavior of nullable
  [SPARK-5562] [MLLIB] LDA should handle empty document.
  [SPARK-6747] [SQL] Throw an AnalysisException when unsupported Java list types used in Hive UDF
  Revert "[SPARK-8781] Fix variables in published pom.xml are not resolved"
  [SPARK-8819] Fix build for maven 3.3.x
  [SPARK-8463][SQL] Use DriverRegistry to load jdbc driver at writing path
  [SPARK-8072] [SQL] Better AnalysisException for writing DataFrame with identically named columns
  [SPARK-8588] [SQL] Regression test
  [SPARK-8765] [MLLIB] Fix PySpark PowerIterationClustering test issue
  Revert "[SPARK-7212] [MLLIB] Add sequence learning flag"
  [SPARK-6707] [CORE] [MESOS] Mesos Scheduler should allow the user to specify constraints based on slave attributes
  ...

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #36744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36744/consoleFull) for   PR 7099 at commit [`3367489`](https://github.com/apache/spark/commit/3367489a1b9b51ed6a76eeb01c68526710ae1f10).

SparkQA commented:
[Test build #36744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36744/console) for   PR 7099 at commit [`3367489`](https://github.com/apache/spark/commit/3367489a1b9b51ed6a76eeb01c68526710ae1f10).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

Commit:
Merge branch 'master' into SPARK-8538

* master:
  [SPARK-7785] [MLLIB] [PYSPARK] Add __str__ and __repr__ to Matrices
  [SPARK-8900] [SPARKR] Fix sparkPackages in init documentation
  [SPARK-8657] [YARN] Fail to upload resource to viewfs
  [SPARK-8888][SQL] Use java.util.HashMap in DynamicPartitionWriterContainer.
  [SPARK-8753][SQL] Create an IntervalType data type
  [SPARK-5707] [SQL] fix serialization of generated projection
  [SPARK-6912] [SQL] Throw an AnalysisException when unsupported Java Map<K,V> types used in Hive UDF
  [SPARK-8785] [SQL] Improve Parquet schema merging
  [SPARK-8894] [SPARKR] [DOC] Example code errors in SparkR documentation.
  [SPARK-8872] [MLLIB] added verification results from R for FPGrowthSuite
  [SPARK-7050] [BUILD] Fix Python Kafka test assembly jar not found issue under Maven build
  [SPARK-8883][SQL]Remove the OverrideFunctionRegistry
  [SPARK-8886][Documentation]python Style update
  [SPARK-8879][SQL] Remove EmptyRow class.
  [SPARK-8878][SQL] Improve unit test coverage for bitwise expressions.
  [SPARK-8868] SqlSerializer2 can go into infinite loop when row consists only of NullType columns
  [SPARK-7190] [SPARK-8804] [SPARK-7815] [SQL] unsafe UTF8String
  [SPARK-8876][SQL] Remove InternalRow type alias in expressions package.
  [SPARK-8794] [SQL] Make PrunedScan work for Sample
  [SPARK-8845] [ML] ML use of Breeze optimization: use adjustedValue  instead of value

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36838/consoleFull) for   PR 7099 at commit [`e71102d`](https://github.com/apache/spark/commit/e71102d1b35ddf8b36c8c63e4e8ca1b2420aaa0b).

SparkQA commented:
[Test build #36838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36838/console) for   PR 7099 at commit [`e71102d`](https://github.com/apache/spark/commit/e71102d1b35ddf8b36c8c63e4e8ca1b2420aaa0b).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Fix build error

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -209,7 +219,13 @@ class LinearRegression(override val uid: String)
     if (handlePersistence) instances.unpersist()
 
-    copyValues(new LinearRegressionModel(uid, weights, intercept))
+    val model = copyValues(new LinearRegressionModel(uid, weights, intercept))
```
No need to call copyValues since this is in train(), not fit().  (train handles that and setParent)  Could you also please correct the use of copyValues earlier in this method?

feynmanliang replied:
Actually, I need to call copyValues there because `model.transform` used on L223 requires `predictCol` to be set. `JavaLinearRegressionSuite#linearRegressionWithSetters()` will fail otherwise.

jkbradley replied:
Oh OK

feynmanliang replied:
The earlier `copyValues` can be removed however since I'm not using `model.transform` on that one, but I'm inclined to leave it just for consistency.

jkbradley replied:
sounds fine

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -209,7 +219,13 @@ class LinearRegression(override val uid: String)
+      model.transform(dataset).select($(predictionCol), $(labelCol)),
+      $(predictionCol),
+      $(labelCol),
+      objectiveTrace.result())
```
objectiveHistory

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -227,12 +243,139 @@ class LinearRegressionModel private[ml] (
+  }
+
+  def hasSummary: Boolean = trainingSummary.isDefined
+
```
remove newline

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -227,12 +243,139 @@ class LinearRegressionModel private[ml] (
+   * Evaluates the model on a testset.
+   * @param dataset Test dataset to evaluate model on.
+   */
+  def evaluate(dataset: DataFrame): LinearRegressionSummary = {
```
@mengxr recommended removing this for now, pending a decision about the method name.

feynmanliang replied:
Will make private and mark with TODO until resolved

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -227,12 +243,139 @@ class LinearRegressionModel private[ml] (
+    predictions: DataFrame,
+    predictionCol: String,
+    labelCol: String,
+    val objectiveTrace: DataFrame)
```
This may be a problem.  If the goal is to save the model (+ summary) via Java serialization, I'm not sure how DataFrames will behave.  Even though they implement serializable, they should definitely not be used on workers.  We may need to rethink the previous design decision about representing non-scalar stats using DataFrames.

feynmanliang replied:
Converting to `Array[Double]` as discussed offline

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -227,12 +243,139 @@ class LinearRegressionModel private[ml] (
+    val objectiveTrace: DataFrame)
+  extends LinearRegressionSummary(predictions, predictionCol, labelCol) {
+
+  def this(
```
Make private to regression

feynmanliang replied:
OK

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -227,12 +243,139 @@ class LinearRegressionModel private[ml] (
+ * @param objectiveTrace objective function (scaled loss + regularization) at each iteration.
+ */
+@Experimental
+class LinearRegressionTrainingSummary private[ml] (
```
Can the constructor be private to regression?

feynmanliang replied:
OK

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

jkbradley commented:
@feynmanliang Apart from those comments, the main other comment I have is that we should force eager evaluation.

@dbtsai Per your discussion with @mengxr this is being refactored to force eager evaluation for cheap metrics (ones requiring little memory and only 1 pass over the data) at the end of training.  Once we add more expensive metrics, we can add parameters to LinearRegression so that the user can specify which of those expensive metrics to compute and save during training.  We decided it was not worth adding parameters for cheap metrics, but it will be important for more expensive metrics.

SparkQA commented:
[Test build #36841 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36841/consoleFull) for   PR 7099 at commit [`0a5e762`](https://github.com/apache/spark/commit/0a5e7625ed4a718a485af19d978d548ab3e7c7d5).

Commit:
Code review changes

SparkQA commented:
[Test build #36841 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36841/console) for   PR 7099 at commit [`0a5e762`](https://github.com/apache/spark/commit/0a5e7625ed4a718a485af19d978d548ab3e7c7d5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

Commit:
Change Summary to hold values, move transient annotations down to metrics and predictions DF

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36854 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36854/consoleFull) for   PR 7099 at commit [`4a42680`](https://github.com/apache/spark/commit/4a42680afdc7f8255f8dbacf5b15a2ab5f4d4ed7).

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -227,16 +243,128 @@ class LinearRegressionModel private[ml] (
+ * @param objectiveHistory objective function (scaled loss + regularization) at each iteration.
+ */
+@Experimental
+class LinearRegressionTrainingSummary private[regression] (
```
Should this be Serializable?

feynmanliang replied:
It inherits Serializable from `LinearRegressionSummary`

jkbradley commented:
That's the only issue I see.  Feel free to fix it in a later PR.  Other than that, LGTM pending tests!

@dbtsai I think these changes will let you save the model with the summary via Java serialization (and we can add it to model.save/load later on).  I'll wait for you to take a final look.

SparkQA commented:
[Test build #36854 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36854/console) for   PR 7099 at commit [`4a42680`](https://github.com/apache/spark/commit/4a42680afdc7f8255f8dbacf5b15a2ab5f4d4ed7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -227,16 +243,128 @@ class LinearRegressionModel private[ml] (
+
+  /**
+   * Gets results summary (e.g. residuals, mse, r-squared ) of model on training set. This method
+   * should only be called on the driver as `summary` is transient.
```
Doc out of date after recent changes.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -227,16 +243,128 @@ class LinearRegressionModel private[ml] (
+    this
+  }
+
+  def hasSummary: Boolean = trainingSummary.isDefined
```
Add doc, if only a reference to summary

jkbradley commented:
Those doc issues could be handled in a follow-up PR if you like.  I'll merge this soon.

Commit:
Update docs

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36973 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36973/consoleFull) for   PR 7099 at commit [`d219fa4`](https://github.com/apache/spark/commit/d219fa4c216e8f35b71a26921561104d15cd6055).

SparkQA commented:
[Test build #36973 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36973/console) for   PR 7099 at commit [`d219fa4`](https://github.com/apache/spark/commit/d219fa4c216e8f35b71a26921561104d15cd6055).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

jkbradley commented:
LGTM I'll merge with master now.  @dbtsai If you have further comments, please say & we can fix in a follow-up.
@feynmanliang Thanks!
